Added Just-In-Time registration for WCF per call

rhyous opened the pull request:
With this change, we can now have Just-in-time registration in a per call scenario. This allows us to register per call objects, such as OperationContext and WebOperationContext.

rhyous committed:
updated version to match latest NuGet package

rhyous committed:
Added Just in time (JIT) module registration so per call settings can be registered.

rhyous committed:
Whitespace

rhyous committed:
Versioned

rhyous committed:
Tagged AutofacInstanceContext

rhyous committed:
Versioned

rhyous committed:
Version restored to original

rhyous committed:
Version restored to original

rhyous committed:
Merge

rhyous committed:
Removed custom build event

rhyous committed:
Removed custom build event

rhyous committed:
Merge

tillig commented on src/Autofac.Integration.Wcf/AutofacInstanceContext.cs:
```diff
@@ -103,7 +105,17 @@ public AutofacInstanceContext(ILifetimeScope container)
                 throw new ArgumentNullException("container");
             }
-            this.OperationLifetime = container.BeginLifetimeScope();
+            this.OperationLifetime = container.BeginLifetimeScope("WcfRequest", (builder) =>
```
The name of the lifetime scope should be a constant rather than an inline magic string so folks can reuse it when registering things. The constants class name [in core Autofac is `MatchingScopeLifetimeTags`](https://github.com/autofac/Autofac/blob/ac840a25dd82b0db3429cd15403411e963bb2102/src/Autofac/Core/Lifetime/MatchingScopeLifetimeTags.cs).

tillig commented on src/Autofac.Integration.Wcf/IPerInstanceContextJitModuleContainer.cs:
```diff
@@ -0,0 +1,19 @@
+    /// static per InstanceContext to be registered and usable throughout
+    /// the instance context.
+    /// </summary>
+    public interface IPerInstanceContextJitModuleContainer
```
There's a general pattern in Autofac of calling these sorts of things "Accessor" - `IPerInstanceContextModuleAccessor` would be good. We know it's per-context so JIT is a little redundant in the name, and "accessor" will fit with the other things named as such.

tillig commented on src/Autofac.Integration.Wcf/IPerInstanceContextJitModuleContainer.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    /// <summary>
+    /// A list of just in time (JIT) registrations. This allows for the
+    /// current OperationContext or WebOperationContext or any item that is
```
Please use `<see cref="..." />` to hyperlink references to other classes/properties - `OperationContext`, `WebOperationContext`, `InstanceContext`.

tillig commented on src/Autofac.Integration.Wcf/IPerInstanceContextJitModuleContainer.cs:
```diff
@@ -0,0 +1,19 @@
+    public interface IPerInstanceContextJitModuleContainer
+    {
+        /// <summary>
+        /// The list of Jit Modules.
```
"Gets the list of per-instance-context modules to register."

tillig commented on test/Autofac.Integration.Wcf.Test/AutofacInstanceContextFixture.cs:
```diff
@@ -63,6 +64,43 @@ public void Resolve_RequiresServiceImplementationData()
             Assert.Throws<ArgumentNullException>(() => context.Resolve(null));
         }
 
+        [Fact]
```
The code in the class handles `.Modules` being null; it might be good to have a test showing that's handled.

tillig commented on src/Autofac.Integration.Wcf/AutofacInstanceContext.cs:
```diff
@@ -103,7 +105,17 @@ public AutofacInstanceContext(ILifetimeScope container)
-            this.OperationLifetime = container.BeginLifetimeScope();
+            this.OperationLifetime = container.BeginLifetimeScope("WcfRequest", (builder) =>
+            {
+                var jitModules = container.ResolveOptional<IPerInstanceContextJitModuleContainer>();
```
This is a potential hit for every request though this isn't the "80% use case." Unfortunately we get hit with complaints about performance at every turn. Have you looked at how this impacts performance, e.g., with some benchmarks before/after?

I'm curious if we may need to, instead, add some sort of "enable this feature" flag somehow so you have to turn it on before it goes through this process. We do that for things like attribute filtering on objects because that's an unnecessary hit if folks don't use it.

Of course, if the hit isn't that bad, maybe it's not needed.

tillig commented on src/Autofac.Integration.Wcf/Autofac.Integration.Wcf.nuspec:
```diff
@@ -4,7 +4,7 @@
     <version>$version$</version>
     <authors>Autofac Contributors</authors>
-    <licenseUrl>https://opensource.org/licenses/mit-license.php</licenseUrl>
+    <license type="file">LICENSE</license>
```
Use `<license type="expression">MIT</license>` - this is [how we do it in core Autofac, too](https://github.com/autofac/Autofac/blob/83ab65cf754507716d4d577e171072c1eab05625/src/Autofac/Autofac.csproj#L21)

tillig reviewed (changes requested):
This looks pretty good but we should definitely verify what the per-call perf hit is to see whether we need to somehow "feature flag" this to avoid the hit if it's measurable.

rhyous committed:
Implement pull requestion suggestions from https://github.com/autofac/Autofac.Wcf/pull/19
1. Avoid possible performance issue by using a feature flag.
2. Make the lifetime scope name a constant in a constants class named MatchingScopeLifetimeTags.cs.
3. Rename IPerInstanceContextJitModuleContainer to IPerInstanceContextModuleAccessor
4. Add cref to comments
5. Add a unit test that verify nulls
6. Update NuSpec for license element.

rhyous commented:
Let me see if I have the suggestion list correct:

1. Avoid possible performance issue by using a feature flag. 
    - For those not using this feature, the performance hit is:
        1. A call to ResolveOptional that returns null.
        2. A null check
    - In order to use a Feature flag, I could create an IFeatureFlags and a concrete, and then try to ResolveOptional that. However, that adds a ResolveOptional hit and null check. So that type of feature flag would result in the same performance hit.
   - I could use a simple static bool as the feature flag. But then all unit tests would share that boolean state. That could be acceptable as all unit tests still pass and the one line not tested is just a simple container.BeginLifetimeScope() call.
   - Done (using the simple static option)
2. Make the lifetime scope name a constant in a constants class named MatchingScopeLifetimeTags.cs. 
   - Done.
3. Rename IPerInstanceContextJitModuleContainer to IPerInstanceContextModuleAccessor 
    - Done
4. Add cref to comments
   - Done
5. Add a unit test that verify nulls
   - Done
6. Update NuSpec for license element.
   - Done

tillig commented:
Sorry, perhaps I was too vague. I was hoping to actually measure the performance hit difference using, say, Benchmark.NET, to verify whether a feature flag was interesting or not. If it turned out super negligible, I'd have said just leave it. I don't know what the actual cost of calling ResolveOptional and null check is vs checking a static value. It may not be that much.

However, I think the static value is probably the right way to go since that will no doubt be cheaper.

In order to be able to add other flags like this, I think having a flags enum with the set of flags and a static that allows us to or-together flags would be better than a single static Boolean.

`AutofacHostFactory.Features |= Features.PerInstanceContextModules;`

By default, the `Features` will be `None.`

```c#
[Flags]
public enum Features {
  None = 0,
  InstancePerContextModules = 1
// Next will be 2, then 4, and so on in powers of two.
}
```

The check could be

`if(AutofacHostFactory.Features.HasFlag(Features.PerInstanceContextModules)`

I'd like to get away from any mentions of "JIT" anywhere. I recognize that's where this started, but the premise is the ability to register modules on a per instance context basis. "JIT" will sound to people to be more about compilation than about registration since that's where it's primarily used.

Since the lifetime scope isn't per request, it's per instance context, the naming of that also has to be changed.

I'll add comments in review format shortly.

tillig commented on src/Autofac.Integration.Wcf/AutofacInstanceContext.cs:
```diff
@@ -25,7 +25,9 @@
 using System.Collections.Generic;
+using System.Linq;
 using System.ServiceModel;
+using System.ServiceModel.Web;
```
Why did this `using` become necessary?

rhyous replied:
Apologies. It is not necessary. That was probably added while I was testing various things and not cleaned up.

tillig commented on src/Autofac.Integration.Wcf/MatchingScopeLifetimeTags.cs:
```diff
@@ -0,0 +1,14 @@
+        /// Tag used in setting up per-request lifetime scope registrations
+        /// (e.g., per-HTTP-request or per-API-request).
+        /// </summary>
+        public static readonly object WcfRequestLifetimeScopeTag = "WcfRequest";
```
Let's make this `WcfInstanceContextLifetimeScopeTag` and `WcfInstanceContext`. Since the instance context is based on the life specified in the service, it's not always per-request. Update the comment accordingly, like:

```xml
<summary>
Tag used in setting up per-instance-context lifetime scope registrations. Note the
instance context is based on the service behavior so it may not be per-request.
</summary>
```

tillig replied:
Also note above comment about whether we should tag the scope at all. If scope tagging is removed, this also gets removed.

rhyous replied:
I read the documentation again. I don't think I need the tag. I have an EntityClient that I want to be different per InstanceContext but used for every request down scope of an InstanceContext. If my implementation of IPerInstanceContextModuleAccessor registers EntityClient as SingleInstance(), if I understand correctly, it will only be a a singleton from the current scope and down. If that is correct, it should meet my need.

tillig replied:
You'll want to test that. I think that's right, but it'd be best to verify. That's a simple test.

tillig replied:
In ASP.NET Core they only provide instance-per-lifetime-scope, no "instance per request," and that seems to work fine. You may be able to sub in instance per lifetime scope and get the expected behavior.

tillig commented on src/Autofac.Integration.Wcf/AutofacInstanceContext.cs:
```diff
@@ -103,7 +110,21 @@ public AutofacInstanceContext(ILifetimeScope container)
             }
-            this.OperationLifetime = container.BeginLifetimeScope();
+            this.OperationLifetime = !JustInTimeModuleRegistration
+                ? container.BeginLifetimeScope()
```
Question: Is the fact that one situation is tagged (with modules) but the other is not (without modules) going to cause people a difficult time in troubleshooting? What if someone intentionally does this somewhere?

```c#
builder.RegisterType<Foo>().InstancePerMatchingLifetimeScope(WcfRequestLifetimeScopeTag);
```

If that's not registered as instance-per-context modules, it'll fail... but if some _totally unrelated_ module _is_ registered instance-per-context, suddenly it'll start working.

I'm curious if tagging the lifetime scope is necessary. Is this to support "instance per lifetime context" objects?

If so, we should tag both when there are registrations _and_ when there aren't.

If it doesn't need to be tagged, let's remove that and the constant altogether.

rhyous replied:
I'll remove the constant altogether

rhyous commented:
Ok. Looking at benchmarking now.

rhyous commented:
I temporarily put in two constructors. 1 with the original code and 1 with the new code. I created four methods to benchmark.

1. CreateWithOriginalCode
2. CreateWithModuleRegistrationFalse
3. CreateWithModuleRegistrationTrueNoModules
4. CreateWithModuleRegistrationTrue1Modules

Here is the benchmark code

```
using System;
using System.Collections.Generic;
using Autofac.Core;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace Autofac.Integration.Wcf.Benchmark
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<BenchmarkTest>();
        }

    }

    public interface IExample { }
    public class Example : IExample { }

    public class ModuleContainer : IPerInstanceContextModuleAccessor
    {
        public IEnumerable<IModule> Modules { get { return _Modules; } }
        private IModule[] _Modules = new[] { new Module1() };
    }

    public class Module1 : Module
    {
        protected override void Load(ContainerBuilder builder)
        {
            var example = new Example();
            builder.RegisterInstance(example).As<IExample>();
        }
    }

    public class BenchmarkTest
    {
        public IContainer Container;
        public ILifetimeScope Scope;

        public BenchmarkTest()
        {
            var builder = new ContainerBuilder();
            Container = builder.Build();
            Scope = Container.BeginLifetimeScope(build =>
            {
                build.RegisterType<ModuleContainer>().As<IPerInstanceContextModuleAccessor>();
            });
        }

        [Benchmark]
        public void CreateWithOriginalCode()
        {
            new AutofacInstanceContext(Container);
        }

        [Benchmark]
        public void CreateWithModuleRegistrationFalse()
        {
            new AutofacInstanceContext(Container, false);
        }

        [Benchmark]
        public void CreateWithModuleRegistrationTrueNoModules()
        {
            new AutofacInstanceContext(Container, true);
        }

        [Benchmark]
        public void CreateWithModuleRegistrationTrue1Modules()
        {
            new AutofacInstanceContext(Scope, true);
        }
    }
}
```

Here is the log

[Autofac.Integration.Wcf.Benchmark.BenchmarkTest-20191216-182416.log](https://github.com/autofac/Autofac.Wcf/files/3970975/Autofac.Integration.Wcf.Benchmark.BenchmarkTest-20191216-182416.log)

The key part of the log
```|                                    Method |      Mean |     Error |    StdDev |
|------------------------------------------ |----------:|----------:|----------:|
|                    CreateWithOriginalCode |  2.231 us | 0.0436 us | 0.0466 us |
|         CreateWithModuleRegistrationFalse |  2.303 us | 0.0440 us | 0.0432 us |
| CreateWithModuleRegistrationTrueNoModules |  5.563 us | 0.1081 us | 0.1245 us |
|  CreateWithModuleRegistrationTrue1Modules | 22.106 us | 0.1910 us | 0.1694 us |
```

So the microseconds increased 2.5 times with ResolveOptional with nothing to resolve, so I would say the feature flag is necessary. With 1 module, the microseconds increased 9.9 times.

rhyous committed:
1. Removed boolean feature flag in favor of a Features enum
2. Renamed objecst to remove JIT nomenclature
3. Removed ILifeTimeScope tag and constant

rhyous commented:
@tillig Thank you for all of your help and especially thank you for your time. Let me know if there is anything else I've missed.

tillig reviewed (approved):
🎉 This is great! Thanks!

rhyous commented:
@tillig Your welcome! I love contributing to open source.
What is the process to get this change into a NuGet package?

tillig commented:
Since we've been making updates to the develop branch for Autofac 5 and rarely (never?) have changes to this library, the next release of this package won't happen until Autofac 5 is out. We do not have a timeline or a formal release process, so I can't tell you exactly when that is. Sorry.
